### DISCUSSION TITLE
JAX-based pymdp v2 for speed improvement?

### DISCUSSION THREAD

**Discussion by mthiboust** [infer-actively/pymdp#128] [Ideas]
Hello,
According to this [recent publication](https://www.sciencedirect.com/science/article/pii/S0925231223011888) about the [Cpp-AIF github project](https://github.com/fgregoretti/Cpp-AIF), the main painpoint of **pymdp** is its speed. Looking at pymdp source code, I see that it could greatly benefit from more vectorized code and hardware acceleration. With its Numpy-like API, JAX could be a great choice here. Are there any plans to rewrite this library using JAX? I'm pretty confident that it could bring several orders of magnitude faster computations while keeping the simplicity of python over C++.

**Comment by conorheins** [comment]
That is exactly our thinking as well! We have been working on the JAX backend for quite some time now actually -- see this branch : https://github.com/infer-actively/pymdp/tree/agent_jax

We haven't benchmarked it yet against that CPP paper but that would be nice to do at some point

**Reply by conorheins** [reply]
Note that this new backend allows fitting of active inference parameters to observed data (e.g. behaviour) via frameworks like numpyro, and also opens it up to hybridisation with autodifferentiable programs like deep nets

**Reply by mthiboust** [reply]
Nice, I'll have a look

**Reply by conorheins** [reply]
@mthiboust just to let you know, the JAX backend has now been merged to main, after nearly 4 years in the making :) 

I have done some anecdotal benchmarking against cpp-AIF and the results are quite strong. If we do single agent active inference for a smal generative model (e.g. the epistemic cue chaining demo) on CPU, then cpp-AIF is still faster even than jax pymdp. But if we move platforms (CPU -> GPU) or do multiple agents (which jax handles naturally through batching, whereas cpp-AIF doesn't have native multi-agent parallelization), then JAX pymdp is significantly faster.

**Reply by mthiboust** [reply]
Nice, well done! I'm not that much into active inference right now but thanks for the update and keep up the good work :+1:
